### PR TITLE
Changes to UCM for PG Share

### PR DESCRIPTION
**Pull request by ChrisPenner**
Builds on #4381 and #4388

## Overview

The remainder of encoding/decoding updates required by PG Share.

## Implementation notes

* Adds `quadmapM` (generalizing existing `quadMap`) for mapping over branch types with effects
* Adds `localizeBranchG`, `localizePatchG` generalizing `localizeBranch`, `localizePatch`
* Generalize a ton of localbranch stuff so we can use it with PG local ids.
* Generalize `localPatchToPatch`
* Export a bunch of traversals that probably should've already existed, since they help me run queries in batches in PG.
* Add `From` instances for most Hash types to simplify conversion between them.
* More error types

## Interesting/controversial decisions

Nah

## Test coverage

Only new exports and combinators, that aren't really used in unison, so existing transcripts should be fine.

## Loose ends

Still need to update the max push workers when PG share actually launches, but we want to delay it until then.
https://github.com/unisonweb/unison/pull/4624

### CODE REVIEW

File: unison-cli/src/Unison/Share/Sync.hs [hunk 1]
```diff
@@ -77,8 +77,10 @@ maxSimultaneousPullDownloaders :: Int
 maxSimultaneousPullDownloaders = 5
 
 -- | The maximum number of push workers at a time. Each push worker reads from the database and uploads entities.
+-- Share currently parallelizes on it's own in the backend, and any more than one push worker
+-- just results in serialization conflicts which slow things down.
 maxSimultaneousPushWorkers :: Int
-maxSimultaneousPushWorkers = 5
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/Share/Sync.hs:81]: This is the one change we may wish to delay until the new share is actually deployed.

File: codebase2/codebase-sqlite/U/Codebase/Sqlite/LocalizeObject.hs [hunk 2]
```diff
@@ -62,70 +68,92 @@ import Unison.Util.Set qualified as Set
...
     localizeTypeEdit =
       bitraverse localizeText localizeDefn
 
 --------------------------------------------------------------------------------------------------------------------------------------------
 -- General-purpose localization
 
 -- Contains references to branch objects.
-type ContainsBranches s =
-  HasType (Map (BranchObjectId, CausalHashId) LocalBranchChildId) s
+class Ord c => ContainsBranches c s where
```
Comment (ChrisPenner) [inline codebase2/codebase-sqlite/U/Codebase/Sqlite/LocalizeObject.hs:112]: Need typeclasses here in order to generalize, but these don't escape the module.

File: codebase2/codebase-sqlite/U/Codebase/Sqlite/LocalizeObject.hs [hunk 3]
```diff
@@ -62,70 +68,92 @@ import Unison.Util.Set qualified as Set
...
+  texts_ :: Lens' s (Map t LocalTextId)
 
 -- The inner state of the localization of a branch object.
-type LocalizeBranchState =
-  ( Map TextId LocalTextId,
-    Map ObjectId LocalDefnId,
-    Map PatchObjectId LocalPatchObjectId,
-    Map (BranchObjectId, CausalHashId) LocalBranchChildId
-  )
+data LocalizeBranchState t d p c = LocalizeBranchState
```
Comment (ChrisPenner) [inline codebase2/codebase-sqlite/U/Codebase/Sqlite/LocalizeObject.hs:132]: Generalize over type-params

File: codebase2/codebase-sqlite/U/Codebase/Sqlite/LocalizeObject.hs (hunk 2 above)
Comment (aryairani) [reply codebase2/codebase-sqlite/U/Codebase/Sqlite/LocalizeObject.hs:112]: Ah, ok.

**Review by aryairani — APPROVED** [review]
@ChrisPenner Feel free to merge when ready

**Pull request merged**